### DISCUSSION TITLE
What gRPC version should I use to build clangd with remote-index enabled?

### DISCUSSION THREAD

**Discussion by LouisHLAV** [clangd/clangd#1303] [Q&A] [answered]
I'm having a hard time building clangd locally with remote-index enabled in RHEL7.
Sadly the binary releases don't work on RHEL7 :(

I keep getting an error on `llvm/llvm-project/clang-tools-extra/clangd/index/remote/monitor/Monitor.cpp:70`:
It seems like the function `Status::error_code()` was [changed](https://github.com/protocolbuffers/protobuf/pull/8354/files) to `Status::code()` in Protobuf v3.16.0.

Any help with the build steps would most welcome as well :)

**Comment by sam-mccall** [comment, marked as answer]
Our releases use grpc 1.36.3 (which looks like protobuf 3.14ish?).
https://github.com/clangd/clangd/blob/master/.github/workflows/autobuild.yaml#L214

That seems pretty old but:
  - it's a version we know works, and we've had multiple problems with buggy versions of grpc
  - debian still only packages 1.30, and we want to stay source-compatible with that

The github actions workflow I linked does fully describe the build on a few OSes (albeit in a clunky syntax), RHEL shouldn't be too different. Any specific problems you're running into?

Sorry the binary releases don't work on RHEL7 - we jump through some hoops to support glibc 2.18 but can't manage earlier versions.
The reason is that on recent systems, some features of the C++ language itself end up depending on glibc 2.18. (Specifically, destructors of thread_locals are implemented using `__cxa_thread_atexit_impl`). To get binaries portable to earlier versions I think we'd have to set up a builder with such an OS, which would be significantly harder to maintain than the ubuntu environments Github provides.

**Reply by LouisHLAV** [reply]
Thank you for the reply, the link and the explanation, @sam-mccall!
I'll try follow this steps as soon as I get access to my RHEL machine again.

After some trial and error, I managed to compile clangd with remote-index support enabled using gRPC 1.49.1.
Sadly, I stopped taking notes after a few tries... All I remember is:
* hitting the issue reported on https://github.com/grpc/grpc/issues/30118 (gcc 9.5 worked fine)
* commenting the `Status` log on `llvm/llvm-project/clang-tools-extra/clangd/index/remote/monitor/Monitor.cpp:70`
* manually compiling `abseil-cpp` and passing the install path to both gRPC's and clangd's cmake calls (see https://github.com/grpc/grpc/issues/27487)
* compiling everything in C++17
* changing some `std::initializer_list`s from `constexpr` to `const` in a particular file (can't really remember which)

Sorry if this is all a little vague, I don't have access to that machine right now and that's all from memory haha
